Source/Fluids/WarpXFluidContainer.cpp

```diff
@@ -0,0 +1,1848 @@
+
+                // Update Q from tn -> tn + dt
+                N_arr(i,j,k) = N_arr(i,j,k)     - (dt/Vij)*(F0_plusx - F0_minusx + F0_plusz - F0_minusz);
+                NUx_arr(i,j,k) = NUx_arr(i,j,k) - (dt/Vij)*(F1_plusx - F1_minusx + F1_plusz - F1_minusz);
+                NUy_arr(i,j,k) = NUy_arr(i,j,k) - (dt/Vij)*(F2_plusx - F2_minusx + F2_plusz - F2_minusz);
+                NUz_arr(i,j,k) = NUz_arr(i,j,k) - (dt/Vij)*(F3_plusx - F3_minusx + F3_plusz - F3_minusz);
+
+#else
+
+                amrex::Real Vz_L_minus = 0.0, Vz_I_minus = 0.0, Vz_L_plus = 0.0, Vz_I_plus = 0.0;
+
+                // Compute the half-timestep velocities
+                if (Q_minus_z(i-1,j,k,0)>0.0) Vz_L_minus = V_calc(Q_minus_z(i-1,j,k,1),Q_minus_z(i-1,j,k,2),Q_minus_z(i-1,j,k,3),clight,2);
+                if (Q_minus_z(i,j,k,0)>0.0)   Vz_I_minus = V_calc(Q_minus_z(i,j,k,1),Q_minus_z(i,j,k,2),Q_minus_z(i,j,k,3),clight,2);
+                if (Q_plus_z(i-1,j,k,0)>0.0)  Vz_L_plus =  V_calc(Q_plus_z(i-1,j,k,1),Q_plus_z(i-1,j,k,2),Q_plus_z(i-1,j,k,3),clight,2);
+                if (Q_plus_z(i,j,k,0)>0.0)    Vz_I_plus =  V_calc(Q_plus_z(i,j,k,1),Q_plus_z(i,j,k,2),Q_plus_z(i,j,k,3),clight,2);
+
+                // compute the fluxes:
+                // (note that _plus is shifted due to grid location)
+                amrex::Real F0_minusz = flux(Q_minus_z(i-1,j,k,0),Q_plus_z(i-1,j,k,0),  Vz_L_minus,Vz_L_plus);
+                amrex::Real F0_plusz =  flux(Q_minus_z(i,j,k,0),  Q_plus_z(i,j,k,0),    Vz_I_minus,Vz_I_plus);
+                amrex::Real F1_minusz = flux(Q_minus_z(i-1,j,k,1)*Q_minus_z(i-1,j,k,0),Q_plus_z(i-1,j,k,1)*Q_plus_z(i-1,j,k,0),  Vz_L_minus,Vz_L_plus);
+                amrex::Real F1_plusz =  flux(Q_minus_z(i,j,k,1)*Q_minus_z(i,j,k,0),  Q_plus_z(i,j,k,1)*Q_plus_z(i,j,k,0),    Vz_I_minus,Vz_I_plus);
+                amrex::Real F2_minusz = flux(Q_minus_z(i-1,j,k,2)*Q_minus_z(i-1,j,k,0),Q_plus_z(i-1,j,k,2)*Q_plus_z(i-1,j,k,0),  Vz_L_minus,Vz_L_plus);
+                amrex::Real F2_plusz =  flux(Q_minus_z(i,j,k,2)*Q_minus_z(i,j,k,0),  Q_plus_z(i,j,k,2)*Q_plus_z(i,j,k,0),    Vz_I_minus,Vz_I_plus);
+                amrex::Real F3_minusz = flux(Q_minus_z(i-1,j,k,3)*Q_minus_z(i-1,j,k,0),Q_plus_z(i-1,j,k,3)*Q_plus_z(i-1,j,k,0),  Vz_L_minus,Vz_L_plus);
+                amrex::Real F3_plusz =  flux(Q_minus_z(i,j,k,3)*Q_minus_z(i,j,k,0),  Q_plus_z(i,j,k,3)*Q_plus_z(i,j,k,0),    Vz_I_minus,Vz_I_plus);
+
+
+                // Update Q from tn -> tn + dt
+                N_arr(i,j,k) = N_arr(i,j,k)     - cz*(F0_plusz - F0_minusz);
+                NUx_arr(i,j,k) = NUx_arr(i,j,k) - cz*(F1_plusz - F1_minusz);
+                NUy_arr(i,j,k) = NUy_arr(i,j,k) - cz*(F2_plusz - F2_minusz);
+                NUz_arr(i,j,k) = NUz_arr(i,j,k) - cz*(F3_plusz - F3_minusz);
+#endif
+            }
+        );
+    }
+}
+
+
+// Momentum source due to curvature
```

Comment:
Should this function between defined and called within an `#ifdef WARPX_DIM_RZ`?

Reply:
The function is only called within an #ifdef WARPX_DIM_RZ in the main evolve loop? Is that what you are asking?

Reply:
Added.